backends/vllm/src/config.rs

```diff
@@ -1,62 +1,254 @@
+    }
+
+    /// Creates a new instance of `ModelConfig` from a `.env` file.
+    pub fn from_env_file() -> Self {
+        dotenv().ok();
+
+        let api_key = std::env::var("API_KEY").ok().unwrap_or_default();
+        let cache_dir = std::env::var("CACHE_DIR")
+            .unwrap_or_default()
+            .parse()
+            .unwrap();
+        let flush_storage = std::env::var("FLUSH_STORAGE")
+            .unwrap_or_default()
+            .parse()
+            .unwrap();
+        let model_name = serde_json::from_str(
+            &std::env::var("MODEL_NAME")
+                .expect("Failed to retrieve models metadata, from .env file"),
+        )
+        .unwrap();
+        let num_tokenizer_workers = serde_json::from_str(
+            &std::env::var("NUM_TOKENIZER_WORKERS")
+                .expect("Failed to retrieve models metadata, from .env file"),
+        )
+        .unwrap();
+        let revision = serde_json::from_str(
+            &std::env::var("REVISION").expect("Failed to retrieve models metadata, from .env file"),
+        )
+        .unwrap();
+        let device_ids = serde_json::from_str(
+            &std::env::var("DEVICE_IDS")
+                .expect("Failed to retrieve models metadata, from .env file"),
+        )
+        .unwrap();
+        let dtype = serde_json::from_str(
+            &std::env::var("DTYPE").expect("Failed to retrieve models metadata, from .env file"),
+        )
+        .unwrap();
+
+        Self {
+            api_key,
+            cache_dir,
+            flush_storage,
+            model_name,
+            num_tokenizer_workers,
+            revision,
+            device_ids,
+            dtype,
+        }
+    }
+}
+
+/// Configuration for the KV cache.
+#[derive(Clone, Debug, Deserialize, Serialize)]
 pub struct CacheConfig {
-    /// Block size
+    /// The size of a block, in number of tokens
     pub(crate) block_size: usize,
-    /// GPU memory utilization
-    gpu_memory_utilization: f32,
-    /// Swap space bytes
-    swap_space_bytes: usize,
-    /// Number of GPU blocks to override (optional)
+    /// The KV cache dtype, if different from the model dtype
+    pub(crate) cache_dtype: Option<String>,
+    /// The fraction of GPU memory to use for the vLLM execution
+    pub(crate) gpu_memory_utilization: f32,
+    /// Fraction of total system memory to use for swap space, this value is
+    /// used to calculate the swap space in bytes if the latter is not specified
+    pub(crate) swap_space_fraction: Option<f32>,
+    /// Size of the CPU swap space per GPU, in bytes, filled using system profiling if not specified
+    pub(crate) swap_space_bytes: Option<usize>,
+    /// Number of GPU blocks to use. This value overrides the profiled `num_gpu_blocks`, if specified
     num_gpu_blocks_override: Option<usize>,
+    /// Total number of GPU blocks to use, filled using system profiling if not specified
+    num_gpu_blocks: Option<usize>,
```

Comment:
Because you are not supposed to fill them in a configuration file, they should be pre-computed by profiling the system's memory, and therefore assigned after the configuration file contents are loaded. Happy to do it in a better way.

Reply:
oh, I see